compiler/rustc_codegen_llvm/src/type_of.rs

```diff
@@ -330,7 +330,9 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyAndLayout<'tcx> {
             ty::Ref(..) | ty::RawPtr(_) => {
                 return self.field(cx, index).llvm_type(cx);
             }
-            ty::Adt(def, _) if def.is_box() => {
+            // only wide pointer boxes are handled as pointers
+            // thin pointer boxes with scalar allocators are handled by the general logic below
+            ty::Adt(def, substs) if def.is_box() && cx.layout_of(substs.type_at(1)).is_zst() => {
```

Comment:
cc @antoyo I think cg_gcc needs the same fix.

Reply:
Seems like applying the same fix cause another [issue](https://github.com/rust-lang/rustc_codegen_gcc/runs/5471850624?check_suite_focus=true).
Do you think there is the same issue for the LLVM codegen?
cc @DrMeepster

Reply:
Looks like the problem is that in the `mini_core` test, Box doesn't have the allocator type parameter.

Reply:
Yeah, fixed it in cg_clif in https://github.com/bjorn3/rustc_codegen_cranelift/commit/401b03443b087e08779878381a170003e2018371. Feel free to copy the changes.

Reply:
Unfortunately, that doesn't solve the issue. I made changes more similar to liballoc itself, but got another ICE:
```
thread 'rustc' panicked at 'assertion failed: sig.c_variadic || extra_args.is_empty()', compiler/rustc_middle/src/ty/layout.rs:3027:13
```

Reply:
Did you change `Box` to include an extra field (eg `()`) **and** change `box_free` to add an extra argument of the same type? They need to be in sync. I got the exact same error when I hadn't changed `box_free` yet.

Reply:
Yes.
What I meant is that I still get the error fixed by this PR with your changes:
```
thread 'rustc' panicked at 'index out of bounds: the len is 1 but the index is 1', /rustc/4b043faba34ccc053a4d0110634c323f6c03765e/compiler/rustc_middle/src/ty/subst.rs:364:43
```

Reply:
Oh, I see what you mean.
My own fixes miss the change you made to `box_free`.
Thanks. It works now!